awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Graph.jsx

```diff
@@ -0,0 +1,756 @@
+    padding-left: 5px;
+  }
+
+  .WorkflowChart-action {
+    height: 25px;
+    width: 25px;
+    font-size: 12px;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    cursor: pointer;
+    border-radius: 2px;
+  }
+
+  .WorkflowChart-action:hover {
+    color: white;
+  }
+
+  .WorkflowChart-action:not(:last-of-type) {
+    margin-bottom: 5px;
+  }
+
+  .WorkflowChart-action--add:hover {
+    background-color: #58B957;
+  }
+
+  .WorkflowChart-action--edit:hover,.WorkflowChart-action--link:hover,.WorkflowChart-action--details:hover {
+    background-color: #0279BC;
+  }
+
+  .WorkflowChart-action--delete:hover {
+    background-color: #D9534F;
+  }
+
+  .WorkflowChart-tooltipArrows {
+    width 10px;
+  }
+
+  .WorkflowChart-tooltipArrows--outer {
+    position:absolute;
+    top: calc(50% - 10px);
+    width: 0;
+    height: 0;
+    border-right: 10px solid #C4C4C4;
+    border-top: 10px solid transparent;
+    border-bottom: 10px solid transparent;
+    margin: auto;
+  }
+
+  .WorkflowChart-tooltipArrows--inner {
+    position:absolute;
+    top: calc(50% - 10px);
+    left: 6px;
+    width: 0;
+    height: 0;
+    border-right: 10px solid white;
+    border-top: 10px solid transparent;
+    border-bottom: 10px solid transparent;
+    margin: auto;
+  }
+
+  .WorkflowChart-tooltipActions {
+    background-color: white;
+    border: 1px solid #C4C4C4;
+    border-radius: 2px;
+    padding: 5px;
+  }
+
+  .WorkflowChart-tooltipContents {
+    display: flex;
+  }
+`;
+
+function Graph({ links, nodes, readOnly, i18n }) {
+  const [helpText, setHelpText] = useState();
+  const svgRef = useRef(null);
+  const gRef = useRef(null);
+  const nodeW = 180;
+  const nodeH = 60;
+  // This needs to be dynamic bc the text can be different lengths in different languages
+  const rootW = 72;
+  const rootH = 40;
+  let currentScale = 1;
+
+  // Dagre is going to shift the root node around as nodes are added/removed
+  // This function ensures that the user doesn't experience that
+  const normalizeY = (nodePositions, y) => {
+    return y - nodePositions[1].y;
+  };
+
+  // This is the zoom function called by using the mousewheel/click and drag
+  const zoom = () => {
+    const translation = [d3.event.transform.x, d3.event.transform.y];
+    d3.select(gRef.current).attr(
+      'transform',
+      `translate(${translation}) scale(${d3.event.transform.k})`
+    );
+    currentScale = d3.event.transform.k;
+  };
+
+  const zoomRef = d3
+    .zoom()
+    .scaleExtent([0.1, 2])
+    .on('zoom', zoom);
+
+  // Initialize the zoom
+  useEffect(() => {
+    d3.select(svgRef.current).call(zoomRef);
+  }, [zoomRef]);
+
+  // Draw the graph - this will get triggered whenever
+  // nodes or links changes
+  useEffect(() => {
+    const nodePositions = {};
+    const line = d3
+      .line()
+      .x(d => {
+        return d.x;
+      })
+      .y(d => {
+        return d.y;
+      });
+    const getLinkOverlayPoints = d => {
```

Comment:
`getLinkOverlayPoints` and [buildGraphArrays](https://github.com/ansible/awx/pull/5379/files#diff-b53e5ed81664e407a270f377944b7dc9R53) seem like they contain some important and intricate logic. If possible, It would be good to move some of this logic into a function for unit testing.